libvast/test/format/simdjson.cpp

```diff
@@ -0,0 +1,177 @@
+
+} // namespace
+
+FIXTURE_SCOPE(zeek_reader_tests, fixtures::deterministic_actor_system)
+
+TEST(simdjson to data) {
+  auto layout = record_type{{"b", bool_type{}},
+                            {"c", count_type{}},
+                            {"r", real_type{}},
+                            {"i", integer_type{}},
+                            {"s", string_type{}},
+                            {"a", address_type{}},
+                            {"sn", subnet_type{}},
+                            {"t", time_type{}},
+                            {"d", duration_type{}},
+                            {"d2", duration_type{}},
+                            {"e", enumeration_type{{"FOO", "BAR", "BAZ"}}},
+                            {"lc", list_type{count_type{}}},
+                            {"lt", list_type{time_type{}}},
+                            {"rec", record_type{{"c", count_type{}},
+                                                {"s", string_type{}}}},
+                            {"msa", map_type{string_type{}, address_type{}}},
+                            {"mcs", map_type{count_type{}, string_type{}}}}
+                  .name("layout");
+  auto flat = flatten(layout);
+  auto builder = factory<table_slice_builder>::make(
+    defaults::import::table_slice_type, flat);
+  std::string_view str = R"json({
+    "b": true,
+    "c": 424242,
+    "r": 4.2,
+    "i": -1337,
+    "s": "0123456789\r\n",
+    "a": "147.32.84.165",
+    "sn": "192.168.0.1/24",
+    "t": "2011-08-12+14:59:11.994970",
+    "d": "42s",
+    "d2": 3.006088,
+    "e": "BAZ",
+    "lc": [ 19, 5555, 0 ],
+    "lt": [ 1556624773, "2019-04-30T11:46:13Z" ],
+    "rec": { "c": 421, "s":"test" },
+    "msa": { "foo": "1.2.3.4", "bar": "2001:db8::" },
+    "mcs": { "1": "FOO", "1024": "BAR!" }
+  })json";
+
+  ::simdjson::dom::parser p;
+  auto el = p.parse(str);
+  CHECK(el.error() == ::simdjson::SUCCESS);
+  auto obj = el.value().get_object();
+  CHECK(obj.error() == ::simdjson::SUCCESS);
+  format::simdjson::add(*builder, obj.value(), flat);
+  auto slice = builder->finish();
+  REQUIRE_NOT_EQUAL(slice.encoding(), table_slice::encoding::none);
+  CHECK(slice.at(0, 0) == data{true});
+  CHECK(slice.at(0, 1) == data{count{424242}});
+
+  CHECK(slice.at(0, 2).is<real>());
+  const auto r = slice.at(0, 2).get_data().get(
+    std::integral_constant<
+      int, caf::detail::tl_index_of<data::types, real>::value>());
+  CHECK(std::abs(r - real{4.2}) < 0.000001);
+
+  CHECK(slice.at(0, 3) == data{integer{-1337}});
+  CHECK_EQUAL(slice.at(0, 4),data{std::string{"0123456789\r\n"}});
```

Comment:
There is a mismatch with previous json behavior. Simdjson handles \r\n and provides data in non-escaped form, a "\r\n" json-string (size 4 characters) becomes a string of size 2 characters ( 0x0D, ox0A). And that breaks some integrations tests, and maybe similar mismatches causes other tests to fail (currently figuring out).

Reply:
I think that's okay. Don't spend too much time on this.

Reply:
Well, it has another catch, export does no escaping and a single record in export results into 2 lines with string containing it  (node-suricata-alert)

Reply:
No escaping for csv-export. Ascii-export is ok

Reply:
That's by design I think (cc @tenzir/backend), csv should not do escaping of newlines.

Reply:
> csv should not do escaping of newlines

The newline is a control character in CSV that gets you to the next record. So when it comes up in strings, it should be escaped. (I'm not sure if that's what you mean.)

Reply:
If you're certain on that front, then we do have a bug here. Let's tackle it independently from this PR, though.

Reply:
I think the naive assumption is that one line is one CSV record. However, everybody does CSV differently. The [Wikipedia page](https://en.wikipedia.org/wiki/Comma-separated_values#General_functionality) sums it up well:

>  Simple CSV implementations may prohibit field values that contain a comma or other special characters such as newlines. More sophisticated CSV implementations permit them, often by requiring " (double quote) characters around values that contain reserved characters (such as commas, double quotes, or less commonly, newlines). Embedded double quote characters may then be represented by a pair of consecutive double quotes,[15] or by prefixing a double quote with an escape character such as a backslash (for example in Sybase Central).

So I wouldn't be surprised if some CSV implementation out there does stateful stuff, such as allowing newlines after an opening double quote or so. I think we should try to do things simple and escape, so that we have one-record-is-one-line semantics. If we ever find a user who needs different semantics, we can still add an option to the command line to change that behavior, but let's start with KISS.